src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/String/KeyAnalyzer.cs

```diff
@@ -296,5 +318,11 @@ private sealed class JustifiedCaseInsensitiveSubstringComparer : SubstringCompar
             public override bool Equals(string? x, string? y) => x.AsSpan(IsLeft ? Index : (x!.Length + Index), Count).Equals(y.AsSpan(IsLeft ? Index : (y!.Length + Index), Count), StringComparison.OrdinalIgnoreCase);
             public override int GetHashCode(string s) => Hashing.GetHashCodeOrdinalIgnoreCase(s.AsSpan(IsLeft ? Index : (s.Length + Index), Count));
         }
+
+        private sealed class JustifiedCaseInsensitiveAsciiSubstringComparer : SubstringComparer
+        {
+            public override bool Equals(string? x, string? y) => x.AsSpan(IsLeft ? Index : (x!.Length + Index), Count).Equals(y.AsSpan(IsLeft ? Index : (y!.Length + Index), Count), StringComparison.OrdinalIgnoreCase);
```

Comment:
i'm curious how these mutable comparers work.  What mutates them, and what do the mutated values mean?

Reply:
ah, i see what hte mutated values mean now.  nut i'm curious where they're set.

Reply:
The code tries to find the smallest substring within each of the keys that sufficiently differentiates the inputs such that hashing just based on that substring results in a number of collisions below a certain small threshold. Rather than creating a new HashSet for each try, it creates a single HashSet with one of these mutable comparers, updating the comparer to the new target region of the strings between tries.

Reply:
> ah, i see what hte mutated values mean now. nut i'm curious where they're set.

e.g. https://github.com/dotnet/runtime/blob/c9bb3e7d203a2bb5469b81653b10a9460ed459d2/src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/String/KeyAnalyzer.cs#L99